### ISSUE TITLE
Custom Decoder For Autoencoder methods

### ISSUE THREAD

**Issue by jlevy44** [rusty1s/pytorch_geometric#220]
Can you add-on the ability for all AE models (maybe just update the AE class) to support user-supplied decoders (which can be problematic for decode_indices)? I'm getting poor results right now just trying to overfit to my original graph (I want my distribution of predictioned values for edges to separate so I can set a threshold to declare 0 or 1), and tried to devise this model, however, decode_indices does not work for reconstruction loss:

```
class VGAE2(VGAE):
    def __init__(self, encoder, p=2):
        super(VGAE2,self).__init__(encoder)
        self.latent_transform = nn.Linear(self.encoder.n_latent,self.encoder.n_latent)
        self.distance_operator = nn.PairwiseDistance(p=p)
        
        
    def decode(self, z, sigmoid=True, pdist=True):
        if pdist:
            print(z)
            adj = torch.exp(-self.distance_operator(z,z))
            print(adj)
        else:
            adj = torch.matmul(z, self.latent_transform(z).t())
        return torch.sigmoid(adj) if sigmoid else adj
```

**Comment by rusty1s** [comment]
Hi, I am not sure if I understand your problem correctly. For using your own decoder, you need to override both `decode` and `decode_indices` for dense and sparse decodings respectively. I do not see a reason that prevents your decoder from implementing `decode_indices`.

**Comment by jlevy44** [comment]
Sure. Why hack it through class inheritance? Wouldn't it be better for the user to input an instance of some decoder class that could serve as one of the input arguments to the GAE class, like the encoders and discriminators? Perhaps if there were a set of decoder classes from pytorch geometric that the user could select from?

**Comment by rusty1s** [comment]
This is a reasonable request. Working on it!

**Comment by rusty1s** [comment]
You can now pass a user-defined decoder module to the Autoencoder models. I'm closing this for now. Feel free to reopen if you encounter any issues.
